packages/api/src/hooks/useCreateActivityStatusRenderer.tsx

```diff
@@ -1,104 +1,107 @@
 /* eslint react/prop-types: "off" */
 /* eslint react/require-default-props: "off" */
 
-import { Constants } from 'botframework-webchat-core';
 import PropTypes from 'prop-types';
-import React, { ReactNode, useMemo, VFC } from 'react';
-import type { WebChatActivity } from 'botframework-webchat-core';
+import React, { memo, useMemo } from 'react';
 
-import useGetSendTimeoutForActivity from './useGetSendTimeoutForActivity';
-import useTimePassed from './internal/useTimePassed';
+import useGetKeyByActivity from '../hooks/useGetKeyByActivity';
+import useSendStatusByActivityKey from '../hooks/useSendStatusByActivityKey';
 import useWebChatAPIContext from './internal/useWebChatAPIContext';
 
-const {
-  ActivityClientState: { SEND_FAILED, SENDING, SENT }
-} = Constants;
+import type { ReactNode, VFC } from 'react';
+import type { RenderActivityStatus } from '../types/ActivityStatusMiddleware';
+import type { SendStatus } from '../providers/ActivitySendStatus/SendStatus';
+import type { WebChatActivity } from 'botframework-webchat-core';
 
-type ActivityStatusContainerProps = {
+type InnerActivityStatusContainerProps = {
   activity: WebChatActivity;
   hideTimestamp: boolean;
   nextVisibleActivity: WebChatActivity;
+  sendStatus: SendStatus;
 };
 
-const ActivityStatusContainer: VFC<ActivityStatusContainerProps> = ({
-  activity,
-  hideTimestamp,
-  nextVisibleActivity
-}) => {
-  const { activityStatusRenderer: createActivityStatusRenderer } = useWebChatAPIContext();
-  const getSendTimeoutForActivity = useGetSendTimeoutForActivity();
-
-  // SEND_FAILED from the activity is ignored, and is instead based on styleOptions.sendTimeout.
-  // Note that the derived state is time-sensitive. The useTimePassed() hook is used to make sure it changes over time.
-  const {
-    from: { role }
-  }: WebChatActivity = activity;
-
-  const fromUser = role === 'user';
-  let activitySent: boolean;
-  let sendTimeoutAt: number;
-
-  if (fromUser) {
-    const state = activity.channelData?.state;
-    const sendTimeout = getSendTimeoutForActivity({ activity });
-
-    activitySent = state !== SENDING && state !== SEND_FAILED;
-
-    // If no timestamp, we assume the "sending" will be timed out as "send failed".
-    sendTimeoutAt = !activitySent
-      ? new Date(activity.localTimestamp || new Date(0).toISOString()).getTime() + sendTimeout
-      : 0;
-  } else {
-    activitySent = true;
-    sendTimeoutAt = 0;
+const ActivityStatusContainerCore: VFC<InnerActivityStatusContainerProps> = memo(
+  ({ activity, hideTimestamp, nextVisibleActivity, sendStatus }) => {
+    const { activityStatusRenderer: createActivityStatusRenderer }: { activityStatusRenderer: RenderActivityStatus } =
```

Comment:
Do we need to specify the type here? Is `useWebChatAPIContext` untyped?

Reply:
Trying to type out `useWebChatAPIContext` and it break something else.

The water bucket is leaking... I will tackle it in a new issue #4367.